README.md
# AutoValue: Ignore Hash Equals Extension [![Build Status](https://travis-ci.org/REggar/auto-value-ignore-hash-equals.svg?branch=master)](https://travis-ci.org/REggar/auto-value-ignore-hash-equals)
An extension for Google's [AutoValue](https://github.com/google/auto/tree/master/value) that omits
`@IgnoreHashEquals` field values from `hashCode()` and `equals()`.


## Basic Usage

Include the extension in your project, define an `@IgnoreHashEquals` annotation, and apply it to any
fields that you wish to be ignored from the generated `hashCode` and `equals` implementation.

```java
@Retention(SOURCE)
@Target({METHOD, PARAMETER, FIELD})
public @interface IgnoreHashEquals {
}
```

```java
@AutoValue
public abstract class User {
  public abstract String id();
  @IgnoreHashEquals public abstract String name();
}
```

When you call `hashCode()` or `equals()` any properties **with** `@IgnoreHashEquals` will be ignored
from the calculation.

## @IncludeHashEquals Usage

Include the extension in your project, define an `@IncludeHashEquals` annotation, and apply it to any
fields that you wish to be included from the generated `hashCode` and `equals` implementation.

```java
@Retention(SOURCE)
@Target({METHOD, PARAMETER, FIELD})
public @interface IncludeHashEquals {
}
```

```java
@AutoValue
public abstract class User {
  @IncludeHashEquals public abstract String id();
  public abstract String name();
}
```

When you call `hashCode()` or `equals()` any properties **without** `@IncludeHashEquals` will be ignored
from the calculation.

## Download

Add a Gradle dependency:

```groovy
apt 'com.github.reggar:auto-value-ignore-hash-equals:1.1.0'
```
(Using the [android-apt](https://bitbucket.org/hvisser/android-apt) plugin)

or Maven:
```xml
<dependency>
  <groupId>com.github.reggar</groupId>
  <artifactId>auto-value-ignore-hash-equals</artifactId>
  <version>1.1.0</version>
  <scope>provided</scope>
</dependency>
```

Snapshots of the development version are available in [Sonatype's `snapshots` repository](https://oss.sonatype.org/content/repositories/snapshots/).


## Notes

This library is heavily inspired by Square's [AutoValue: Redacted Extension](https://github.com/square/auto-value-redacted).


## License

```
Copyright 2016 Robert Eggar.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```

CHANGELOG.md
Change Log
==========

Version 1.0.0 *(2016-10-19)*
----------------------------
Initial release.


src/test/java/com/github/reggar/ignorehashequals/AnnotationTypeTest.java
package com.github.reggar.ignorehashequals;

import com.google.common.collect.Sets;

import org.junit.Test;

import java.util.Collections;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class AnnotationTypeTest {

    @Test public void shouldBeIncluded() throws Exception {
        assertTrue(AnnotationType.INCLUDE.shouldBeIncluded(Collections.singleton("IncludeHashEquals")));
        assertFalse(AnnotationType.IGNORE.shouldBeIncluded(Collections.singleton("IgnoreHashEquals")));

        assertTrue(AnnotationType.IGNORE.shouldBeIncluded(Collections.singleton("IncludeHashEquals")));
        assertFalse(AnnotationType.INCLUDE.shouldBeIncluded(Collections.singleton("IgnoreHashEquals")));

        assertTrue(AnnotationType.IGNORE.shouldBeIncluded(Collections.singleton("SomethingElse")));
        assertFalse(AnnotationType.INCLUDE.shouldBeIncluded(Collections.singleton("SomethingElse")));

        assertTrue(AnnotationType.NOT_PRESENT.shouldBeIncluded(Collections.<String>emptySet()));
        assertTrue(AnnotationType.ERROR.shouldBeIncluded(Collections.<String>emptySet()));
    }

    @Test public void from() throws Exception {
        assertEquals(AnnotationType.INCLUDE, AnnotationType.from(Collections.singleton("IncludeHashEquals")));

        assertEquals(AnnotationType.IGNORE, AnnotationType.from(Collections.singleton("IgnoreHashEquals")));

        assertEquals(AnnotationType.NOT_PRESENT, AnnotationType.from(Collections.<String>emptySet()));

        assertEquals(AnnotationType.NOT_PRESENT, AnnotationType.from(Collections.singleton("SomethingElse")));

        assertEquals(AnnotationType.ERROR, AnnotationType.from(Sets.newHashSet("IncludeHashEquals", "IgnoreHashEquals")));
    }
}

src/test/java/com/github/reggar/ignorehashequals/AutoValueIgnoreHashEqualsExtensionTest.java
/*
 * Copyright (C) 2016 Robert Eggar.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.reggar.ignorehashequals;

import com.google.auto.value.processor.AutoValueProcessor;
import com.google.testing.compile.JavaFileObjects;
import java.util.Arrays;
import javax.tools.JavaFileObject;
import org.junit.Before;
import org.junit.Test;

import static com.google.common.truth.Truth.assertAbout;
import static com.google.testing.compile.JavaSourcesSubjectFactory.javaSources;

public final class AutoValueIgnoreHashEqualsExtensionTest {
  private JavaFileObject ignoreHashEquals;
  private JavaFileObject includeHashEquals;
  private JavaFileObject nullable;

  @Before public void setUp() {
    ignoreHashEquals = JavaFileObjects.forSourceString("test.IgnoreHashEquals", ""
        + "package test;\n"
        + "import java.lang.annotation.Retention;\n"
        + "import java.lang.annotation.Target;\n"
        + "import static java.lang.annotation.ElementType.FIELD;\n"
        + "import static java.lang.annotation.ElementType.METHOD;\n"
        + "import static java.lang.annotation.ElementType.PARAMETER;\n"
        + "import static java.lang.annotation.RetentionPolicy.SOURCE;\n"
        + "@Retention(SOURCE)\n"
        + "@Target({METHOD, PARAMETER, FIELD})\n"
        + "public @interface IgnoreHashEquals {\n"
        + "}");
    includeHashEquals = JavaFileObjects.forSourceString("test.IncludeHashEquals", ""
        + "package test;\n"
        + "import java.lang.annotation.Retention;\n"
        + "import java.lang.annotation.Target;\n"
        + "import static java.lang.annotation.ElementType.FIELD;\n"
        + "import static java.lang.annotation.ElementType.METHOD;\n"
        + "import static java.lang.annotation.ElementType.PARAMETER;\n"
        + "import static java.lang.annotation.RetentionPolicy.SOURCE;\n"
        + "@Retention(SOURCE)\n"
        + "@Target({METHOD, PARAMETER, FIELD})\n"
        + "public @interface IncludeHashEquals {\n"
        + "}");
    nullable = JavaFileObjects.forSourceString("test.Nullable", ""
        + "package test;\n"
        + "import java.lang.annotation.Retention;\n"
        + "import java.lang.annotation.Target;\n"
        + "import static java.lang.annotation.ElementType.FIELD;\n"
        + "import static java.lang.annotation.ElementType.METHOD;\n"
        + "import static java.lang.annotation.ElementType.PARAMETER;\n"
        + "import static java.lang.annotation.RetentionPolicy.CLASS;\n"
        + "@Retention(CLASS)\n"
        + "@Target({METHOD, PARAMETER, FIELD})\n"
        + "public @interface Nullable {\n"
        + "}");
  }

  @Test public void ignoreHashEqualsAnnotation() {
    JavaFileObject source = JavaFileObjects.forSourceString("test.Test", ""
            + "package test;\n"
            + "import com.google.auto.value.AutoValue;\n"
            + "@AutoValue public abstract class Test {\n"
            + "@IgnoreHashEquals public abstract String a();\n"
            + "public abstract int b();\n"
            + "public abstract long c();\n"
            + "public abstract float d();\n"
            + "public abstract double e();\n"
            + "public abstract boolean f();\n"
            + "public abstract int[] g();\n"
            + "public abstract String h();\n"
            + "@Nullable public abstract String i();\n"
            + "}\n"
    );

    JavaFileObject expectedSource = JavaFileObjects.forSourceString("test/AutoValue_Test", ""
                + "package test;\n"
                + "\n"
                + "import java.lang.Object;\n"
                + "import java.lang.Override;\n"
                + "import java.lang.String;\n"
                + "import java.util.Arrays;\n"
                + "\n"
                + "final class AutoValue_Test extends $AutoValue_Test {\n"
                + "  AutoValue_Test(String a, int b, long c, float d, double e, boolean f, int[] g, String h, String i) {\n"
                + "    super(a, b, c, d, e, f, g, h, i);\n"
                + "  }\n"
                + "\n"
                + "  @Override\n"
                + "  public final boolean equals(Object o) {\n"
                + "    if (o == this) {\n"
                + "      return true;\n"
                + "    }\n"
                + "    if (o instanceof Test) {\n"
                + "      Test that = (Test) o;\n"
                + "      return (this.b() == that.b())\n"
                + "          && (this.c() == that.c())\n"
                + "          && (Float.floatToIntBits(this.d()) == Float.floatToIntBits(that.d()))\n"
                + "          && (Double.doubleToLongBits(this.e()) == Double.doubleToLongBits(that.e()))\n"
                + "          && (this.f() == that.f())\n"
                + "          && (Arrays.equals(this.g(), that.g()))\n"
                + "          && (this.h().equals(that.h()))\n"
                + "          && ((this.i() == null) ? (that.i() == null) : this.i().equals(that.i()));\n"
                + "    }\n"
                + "    return false;\n"
                + "  }\n"
                + "\n"
                + "  @Override\n"
                + "  public final int hashCode() {\n"
                + "    int h = 1;\n"
                + "    h *= 1000003;\n"
                + "    h ^= this.b();\n"
                + "    h *= 1000003;\n"
                + "    h ^= (this.c() >>> 32) ^ this.c();\n"
                + "    h *= 1000003;\n"
                + "    h ^= Float.floatToIntBits(this.d());\n"
                + "    h *= 1000003;\n"
                + "    h ^= (Double.doubleToLongBits(this.e()) >>> 32) ^ Double.doubleToLongBits(this.e());\n"
                + "    h *= 1000003;\n"
                + "    h ^= this.f() ? 1231 : 1237;\n"
                + "    h *= 1000003;\n"
                + "    h ^= java.util.Arrays.hashCode(this.g());\n"
                + "    h *= 1000003;\n"
                + "    h ^= this.h().hashCode();\n"
                + "    h *= 1000003;\n"
                + "    h ^= (i() == null) ? 0 : this.i().hashCode();\n"
                + "    return h;\n"
                + "  }\n"
                + "}"
    );

    assertAbout(javaSources())
        .that(Arrays.asList(ignoreHashEquals, nullable, source))
        .processedWith(new AutoValueProcessor())
        .compilesWithoutError()
        .and()
        .generatesSources(expectedSource);
  }

  @Test public void includeHashEqualsAnnotation() {
    JavaFileObject source = JavaFileObjects.forSourceString("test.Test", ""
            + "package test;\n"
            + "import com.google.auto.value.AutoValue;\n"
            + "@AutoValue public abstract class Test {\n"
            + "@IncludeHashEquals @Nullable public abstract String a();\n"
            + "@IncludeHashEquals public abstract int b();\n"
            + "public abstract long c();\n"
            + "public abstract float d();\n"
            + "public abstract double e();\n"
            + "public abstract boolean f();\n"
            + "public abstract int[] g();\n"
            + "public abstract String h();\n"
            + "@Nullable public abstract String i();\n"
            + "}\n"
    );

    JavaFileObject expectedSource = JavaFileObjects.forSourceString("test/AutoValue_Test", ""
            + "package test;\n"
            + "\n"
            + "import java.lang.Object;\n"
            + "import java.lang.Override;\n"
            + "import java.lang.String;\n"
            + "\n"
            + "final class AutoValue_Test extends $AutoValue_Test {\n"
            + "  AutoValue_Test(String a, int b, long c, float d, double e, boolean f, int[] g, String h, String i) {\n"
            + "    super(a, b, c, d, e, f, g, h, i);\n"
            + "  }\n"
            + "\n"
            + "  @Override\n"
            + "  public final boolean equals(Object o) {\n"
            + "    if (o == this) {\n"
            + "      return true;\n"
            + "    }\n"
            + "    if (o instanceof Test) {\n"
            + "      Test that = (Test) o;\n"
            + "      return ((this.a() == null) ? (that.a() == null) : this.a().equals(that.a()))\n"
            + "          && (this.b() == that.b());\n"
            + "    }\n"
            + "    return false;\n"
            + "  }\n"
            + "\n"
            + "  @Override\n"
            + "  public final int hashCode() {\n"
            + "    int h = 1;\n"
            + "    h *= 1000003;\n"
            + "    h ^= (a() == null) ? 0 : this.a().hashCode();\n"
            + "    h *= 1000003;\n"
            + "    h ^= this.b();\n"
            + "    return h;\n"
            + "  }\n"
            + "}"
    );

    assertAbout(javaSources())
            .that(Arrays.asList(includeHashEquals, nullable, source))
            .processedWith(new AutoValueProcessor())
            .compilesWithoutError()
            .and()
            .generatesSources(expectedSource);
  }

  @Test public void annotationsAreMutuallyExclusive() {
    JavaFileObject source = JavaFileObjects.forSourceString("test.Test", ""
            + "package test;\n"
            + "import com.google.auto.value.AutoValue;\n"
            + "@AutoValue public abstract class Test {\n"
            + "@IncludeHashEquals @Nullable public abstract String a();\n"
            + "@IncludeHashEquals public abstract int b();\n"
            + "@IgnoreHashEquals public abstract long c();\n"
            + "public abstract float d();\n"
            + "public abstract double e();\n"
            + "public abstract boolean f();\n"
            + "public abstract int[] g();\n"
            + "public abstract String h();\n"
            + "@Nullable public abstract String i();\n"
            + "}\n"
    );

    assertAbout(javaSources())
            .that(Arrays.asList(includeHashEquals, nullable, source))
            .processedWith(new AutoValueProcessor())
            .failsToCompile();
  }
}


src/test/java/com/github/reggar/includehashequals/AutoValueIncludeHashEqualsExtensionTest.java
/*
 * Copyright (C) 2016 Robert Eggar.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.reggar.includehashequals;

import com.google.auto.value.processor.AutoValueProcessor;
import com.google.testing.compile.JavaFileObjects;

import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;

import javax.tools.JavaFileObject;

import static com.google.common.truth.Truth.assertAbout;
import static com.google.testing.compile.JavaSourcesSubjectFactory.javaSources;

public final class AutoValueIncludeHashEqualsExtensionTest {
    private JavaFileObject nullable;

    @Before
    public void setUp() {
        nullable = JavaFileObjects.forSourceString("test.Nullable", ""
                + "package test;\n"
                + "import java.lang.annotation.Retention;\n"
                + "import java.lang.annotation.Target;\n"
                + "import static java.lang.annotation.ElementType.FIELD;\n"
                + "import static java.lang.annotation.ElementType.METHOD;\n"
                + "import static java.lang.annotation.ElementType.PARAMETER;\n"
                + "import static java.lang.annotation.RetentionPolicy.CLASS;\n"
                + "@Retention(CLASS)\n"
                + "@Target({METHOD, PARAMETER, FIELD})\n"
                + "public @interface Nullable {\n"
                + "}");
    }

    @Test
    public void simple() {
        JavaFileObject source = JavaFileObjects.forSourceString("test.Test", ""
                + "package test;\n"
                + "import com.google.auto.value.AutoValue;\n"
                + "import com.github.reggar.includehashequals.IncludeHashEquals;\n"
                + "@AutoValue public abstract class Test {\n"
                + "@IncludeHashEquals @Nullable public abstract String a();\n"
                + "@IncludeHashEquals public abstract int b();\n"
                + "public abstract long c();\n"
                + "public abstract float d();\n"
                + "public abstract double e();\n"
                + "public abstract boolean f();\n"
                + "public abstract int[] g();\n"
                + "public abstract String h();\n"
                + "@Nullable public abstract String i();\n"
                + "}\n"
        );

        JavaFileObject expectedSource = JavaFileObjects.forSourceString("test/AutoValue_Test", ""
                + "package test;\n"
                + "\n"
                + "import java.lang.Object;\n"
                + "import java.lang.Override;\n"
                + "import java.lang.String;\n"
                + "\n"
                + "final class AutoValue_Test extends $AutoValue_Test {\n"
                + "  AutoValue_Test(String a, int b, long c, float d, double e, boolean f, int[] g, String h, String i) {\n"
                + "    super(a, b, c, d, e, f, g, h, i);\n"
                + "  }\n"
                + "\n"
                + "  @Override\n"
                + "  public final boolean equals(Object o) {\n"
                + "    if (o == this) {\n"
                + "      return true;\n"
                + "    }\n"
                + "    if (o instanceof Test) {\n"
                + "      Test that = (Test) o;\n"
                + "      return ((this.a() == null) ? (that.a() == null) : this.a().equals(that.a()))\n"
                + "          && (this.b() == that.b());\n"
                + "    }\n"
                + "    return false;\n"
                + "  }\n"
                + "\n"
                + "  @Override\n"
                + "  public final int hashCode() {\n"
                + "    int h = 1;\n"
                + "    h *= 1000003;\n"
                + "    h ^= (a() == null) ? 0 : this.a().hashCode();\n"
                + "    h *= 1000003;\n"
                + "    h ^= this.b();\n"
                + "    return h;\n"
                + "  }\n"
                + "}"
        );

        assertAbout(javaSources())
                .that(Arrays.asList(nullable, source))
                .processedWith(new AutoValueProcessor())
                .compilesWithoutError()
                .and()
                .generatesSources(expectedSource);
    }
}


.buildscript/deploy_snapshot.sh
#!/bin/bash
#
# Deploy a jar, source jar, and javadoc jar to Sonatype's snapshot repo.
#
# Adapted from https://coderwall.com/p/9b_lfq and
# http://benlimmer.com/2013/12/26/automatically-publish-javadoc-to-gh-pages-with-travis-ci/

SLUG="REggar/auto-value-ignore-hash-equals"
JDK="oraclejdk8"
BRANCH="master"

set -e

if [ "$TRAVIS_REPO_SLUG" != "$SLUG" ]; then
  echo "Skipping snapshot deployment: wrong repository. Expected '$SLUG' but was '$TRAVIS_REPO_SLUG'."
elif [ "$TRAVIS_JDK_VERSION" != "$JDK" ]; then
  echo "Skipping snapshot deployment: wrong JDK. Expected '$JDK' but was '$TRAVIS_JDK_VERSION'."
elif [ "$TRAVIS_PULL_REQUEST" != "false" ]; then
  echo "Skipping snapshot deployment: was pull request."
elif [ "$TRAVIS_BRANCH" != "$BRANCH" ]; then
  echo "Skipping snapshot deployment: wrong branch. Expected '$BRANCH' but was '$TRAVIS_BRANCH'."
else
  echo "Deploying snapshot..."
  ./gradlew clean uploadArchives
  echo "Snapshot deployed!"
fi


src/main/java/com/github/reggar/ignorehashequals/AnnotationType.java
package com.github.reggar.ignorehashequals;

import java.util.Set;

enum AnnotationType {
    ERROR(""), // Error case
    INCLUDE("IncludeHashEquals"),
    IGNORE("IgnoreHashEquals"),
    NOT_PRESENT(""); // Empty case

    private final String annotationName;

    AnnotationType(String annotationName) {
        this.annotationName = annotationName;
    }

    public boolean shouldBeIncluded(Set<String> propertyAnnotations) {
        boolean annotationPresent = propertyAnnotations.contains(annotationName);
        switch (this) {
            case INCLUDE:
                return annotationPresent;
            case IGNORE:
                return !annotationPresent;
            default:
                return true;
        }
    }

    public static AnnotationType from(Set<String> annotations) {
        boolean ignoreHashEqualsPresent = annotations.contains(IGNORE.annotationName);
        boolean includeHashEqualsPresent = annotations.contains(INCLUDE.annotationName);

        if (ignoreHashEqualsPresent && includeHashEqualsPresent) {
            return ERROR;
        } else if (ignoreHashEqualsPresent) {
            return IGNORE;
        } else if (includeHashEqualsPresent) {
            return INCLUDE;
        } else {
            return NOT_PRESENT;
        }
    }
}


src/main/java/com/github/reggar/ignorehashequals/AutoValueIgnoreHashEqualsExtension.java
/*
 * Copyright (C) 2016 Robert Eggar.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.reggar.ignorehashequals;

import com.google.auto.service.AutoService;
import com.google.auto.value.extension.AutoValueExtension;
import com.squareup.javapoet.ArrayTypeName;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.JavaFile;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeSpec;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.Name;

@AutoService(AutoValueExtension.class) public final class AutoValueIgnoreHashEqualsExtension
    extends AutoValueExtension {

  private static AnnotationType annotationType;

  @Override public boolean applicable(Context context) {
    Map<String, ExecutableElement> properties = context.properties();
    for (ExecutableElement element : properties.values()) {
      Set<String> annotations = getAnnotations(element);

      annotationType = AnnotationType.from(annotations);
      if (annotationType == AnnotationType.ERROR) {
        throw new RuntimeException("Annotations are mutually exclusive, " +
                "only one annotation type can be included at the same time.");
      }

      return annotationType != AnnotationType.NOT_PRESENT;
    }

    return false;
  }

  @Override public String generateClass(Context context, String className, String classToExtend,
      boolean isFinal) {
    String packageName = context.packageName();
    Name superName = context.autoValueClass().getSimpleName();
    Map<String, ExecutableElement> properties = context.properties();

    TypeSpec subclass = TypeSpec.classBuilder(className) //
        .addModifiers(isFinal ? Modifier.FINAL : Modifier.ABSTRACT) //
        .superclass(ClassName.get(packageName, classToExtend)) //
        .addMethod(generateConstructor(properties)) //
        .addMethod(generateEquals(superName, properties)) //
        .addMethod(generateHashCode(superName, properties)) //
        .build();

    JavaFile javaFile = JavaFile.builder(packageName, subclass).build();
    return javaFile.toString();
  }

  private static MethodSpec generateConstructor(Map<String, ExecutableElement> properties) {
    List<ParameterSpec> params = new ArrayList<>();
    for (Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
      TypeName typeName = TypeName.get(entry.getValue().getReturnType());
      params.add(ParameterSpec.builder(typeName, entry.getKey()).build());
    }

    StringBuilder body = new StringBuilder("super(");
    for (int i = properties.size(); i > 0; i--) {
      body.append("$N");
      if (i > 1) body.append(", ");
    }
    body.append(")");

    return MethodSpec.constructorBuilder() //
        .addParameters(params) //
        .addStatement(body.toString(), properties.keySet().toArray()) //
        .build();
  }

  private static MethodSpec generateHashCode(Name superName,
      Map<String, ExecutableElement> properties) {
    MethodSpec.Builder builder = MethodSpec.methodBuilder("hashCode") //
        .addAnnotation(Override.class) //
        .addModifiers(Modifier.PUBLIC, Modifier.FINAL) //
        .returns(TypeName.INT) //
        .addCode("int h = 1;\n");

    for (Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
      ExecutableElement propertyElement = entry.getValue();
      Set<String> propertyAnnotations = getAnnotations(propertyElement);

      if (annotationType.shouldBeIncluded(propertyAnnotations)) {
        builder.addCode("h *= 1000003;\n");
        builder.addCode("h ^= " + generateHashCodeExpression(propertyElement) + ";");
      }

      builder.addCode("\n");
    }

    return builder
        .addCode("return h;\n")
        .build();
  }

  private static CodeBlock generateHashCodeExpression(ExecutableElement propertyElement) {
    String methodName = propertyElement.getSimpleName().toString();
    TypeName propertyType = TypeName.get(propertyElement.getReturnType());
    Set<String> propertyAnnotations = getAnnotations(propertyElement);

    boolean nullable = propertyAnnotations.contains("Nullable");

    if (propertyType.equals(TypeName.BYTE) || propertyType.equals(TypeName.SHORT) ||
        propertyType.equals(TypeName.CHAR) || propertyType.equals(TypeName.INT)) {
      return CodeBlock.of("this.$N()", methodName);
    } else if (propertyType.equals(TypeName.LONG)) {
      return CodeBlock.of("(this.$N() >>> 32) ^ this.$N()", methodName, methodName);
    } else if (propertyType.equals(TypeName.FLOAT)) {
      return CodeBlock.of("Float.floatToIntBits(this.$N())", methodName);
    } else if (propertyType.equals(TypeName.DOUBLE)) {
      return CodeBlock.of(
          "(Double.doubleToLongBits(this.$N()) >>> 32) ^ Double.doubleToLongBits(this.$N())",
          methodName, methodName);
    } else if (propertyType.equals(TypeName.BOOLEAN)) {
      return CodeBlock.of("this.$N() ? 1231 : 1237", methodName);
    } else if (propertyType instanceof ArrayTypeName) {
      return CodeBlock.of("$T.hashCode(this.$N())", Arrays.class, methodName);
    } else {
      if (nullable) {
        return CodeBlock.of("($N() == null) ? 0 : this.$N().hashCode()", methodName, methodName);
      } else {
        return CodeBlock.of("this.$N().hashCode()", methodName);
      }
    }
  }

  private static MethodSpec generateEquals(Name superName,
      Map<String, ExecutableElement> properties) {
    MethodSpec.Builder builder = MethodSpec.methodBuilder("equals") //
        .addAnnotation(Override.class) //
        .addModifiers(Modifier.PUBLIC, Modifier.FINAL) //
        .returns(TypeName.BOOLEAN) //
        .addParameter(TypeName.OBJECT, "o")
        .addCode("if (o == this) {\n")
        .addCode("  return true;\n")
        .addCode("}\n")
        .addCode("if (o instanceof $L) {\n", superName)
        .addCode("  $L that = ($L) o;\n", superName, superName);

    if (properties.size() == 0) {
      builder.addCode("return true;\n");
    } else {
      builder.addCode("  return ");
    }

    List<ExecutableElement> nonIgnoredExecutableElements = new ArrayList<>();

    for(Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
      ExecutableElement propertyElement = entry.getValue();
      Set<String> propertyAnnotations = getAnnotations(propertyElement);

      if(annotationType.shouldBeIncluded(propertyAnnotations)) {
        nonIgnoredExecutableElements.add(propertyElement);
      }
    }

    boolean last = false;
    for (int i = 0; i < nonIgnoredExecutableElements.size(); i++) {
      ExecutableElement propertyElement = nonIgnoredExecutableElements.get(i);

      if(i == nonIgnoredExecutableElements.size() - 1) {
        last = true;
      }

      builder.addCode(generateEqualsExpression(propertyElement));
      if (!last) {
        builder.addCode("\n      && ");
      }
    }

    return builder
        .addCode(";\n")
        .addCode("}\n")
        .addCode("return false;\n")
        .build();
  }

  private static CodeBlock generateEqualsExpression(ExecutableElement propertyElement) {
    String methodName = propertyElement.getSimpleName().toString();
    TypeName propertyType = TypeName.get(propertyElement.getReturnType());
    Set<String> propertyAnnotations = getAnnotations(propertyElement);

    boolean nullable = propertyAnnotations.contains("Nullable");

    if (propertyType.equals(TypeName.FLOAT)) {
      return CodeBlock.of("(Float.floatToIntBits(this.$N()) == Float.floatToIntBits(that.$N()))",
          methodName, methodName);
    } else if (propertyType.equals(TypeName.DOUBLE)) {
      return CodeBlock.of(
          "(Double.doubleToLongBits(this.$N()) == Double.doubleToLongBits(that.$N()))", methodName,
          methodName);
    } else if (propertyType.isPrimitive()) {
      return CodeBlock.of("(this.$N() == that.$N())", methodName, methodName);
    } else if (propertyType instanceof ArrayTypeName) {
      return CodeBlock.of("($T.equals(this.$N(), that.$N()))", Arrays.class, methodName,
          methodName);
    } else {
      if (nullable) {
        return CodeBlock.of(
            "((this.$N() == null) ? (that.$N() == null) : this.$N().equals(that.$N()))", methodName,
            methodName, methodName, methodName);
      } else {
        return CodeBlock.of("(this.$N().equals(that.$N()))", methodName, methodName);
      }
    }
  }

  private static Set<String> getAnnotations(ExecutableElement element) {
    Set<String> set = new LinkedHashSet<>();

    List<? extends AnnotationMirror> annotations = element.getAnnotationMirrors();
    for (AnnotationMirror annotation : annotations) {
      set.add(annotation.getAnnotationType().asElement().getSimpleName().toString());
    }

    return Collections.unmodifiableSet(set);
  }
}
